docs/topic/rmq.md

```diff
@@ -1,115 +1,111 @@
-
-在 ST 表的基础上 Four russian 算法对其做出的改进是序列分块。
+The [Method of Four Russians](https://en.wikipedia.org/wiki/Method_of_Four_Russians) is an algorithm based on ST table proposed by four Russian computer scientists.
 
-具体来说，我们将原数组——我们将其称之为数组 A——每 $S$ 个分成一块，总共 $n/S$ 块。
+The improvement made by the Four Russians algorithm based on the ST table is sequence blocking.
 
-对于每一块我们预处理出来块内元素的最小值，建立一个长度为 $n/S$ 的数组 B，并对数组 B 采用 ST 表的方式预处理。
+Specifically speaking, we divide the original array, let's call it array A, into blocks of every $S$, for a total of $n/S$ blocks.
 
-同时，我们对于数组 A 的每一个零散块也建立一个 ST 表。
+For each block, we precomputate it to get the minimum value of the block, then create an array B with a length of $n/S$, and precomputate the array B using the ST table.
 
-询问的时候，我们可以将询问区间划分为不超过 1 个数组 B 上的连续块区间和不超过 2 个数组 A 上的整块内的连续区间。显然这些问题我们通过 ST 表上的区间查询解决。
+At the same time, we also create a ST table for each fragmented block of array A.
 
-在 $S=\log n$ 时候，预处理复杂度达到最优，为 $O((n / \log n)\log n+(n / \log n)\times\log n\times\log \log n)=O(n\log \log n)$ 。
+When accessing the array, we can divide the query interval into no more than one continuous block interval of array B and no more than two continuous intervals within the entire block of array A. Obviously, these problems can be solved by interval query in the ST table.
 
-时间复杂度 $O(n\log \log n)-O(1)$ 
+When $S=\log n$ , the precomputation complexity reaches the optimal, $O((n / \log n)\log n+(n / \log n)\times\log n\times\log \log n)=O(n\log \log n)$ 。
 
-空间复杂度 $O(n\log \log n)$ 
+Time complexity $O(n\log \log n)-O(1)$ 
 
-当然询问由于要跑三个 ST 表，该实现方法的常数较大。
+Space complexity $O(n\log \log n)$ 
 
-!!! note "一些小小的算法改进"
-    我们发现，在询问的两个端点在数组 A 中属于不同的块的时候，数组 A 中块内的询问是关于每一块前缀或者后缀的询问。
+Of course, because we need to run three ST tables, the constant of this method is relatively large.
 
-    显然这些询问可以通过预处理答案在 $O(n)$ 的时间复杂度内被解决。
+!!! note "Some minor algorithm optimizations"
+    We found that when the two endpoints of the query belong to different blocks in array A, the query in the blocks of array A is about the prefix or suffix of each block.
 
-    这样子我们只需要在询问的时候进行至多一次 ST 表上的查询操作了。
+    Obviously, these queries can be solved within the time complexity of $O(n)$ by preprocessing the answers.
 
-!!! note "一些玄学的算法改进"
-    由于 Four russian 算法以 ST 表为基础，而算法竞赛一般没有非常高的时间复杂度要求，所以 Four russian 算法一般都可以被 ST 表代替，在算法竞赛中并不实用。这里提供一种在算法竞赛中更加实用的 Four russian 改进算法。
+    In this way, we only need to perform the query operation on the ST table at most once.
 
-    我们将块大小设为 $\sqrt n$，然后预处理出每一块内前缀和后缀的 RMQ，再暴力预处理出任意连续的整块之间的 RMQ，时间复杂度为 $O(n)$。
+!!! note "Some algorithm optimizations[ONLY FOR REFERENCE]"
+    Because the Four Russians algorithm is based on the ST table, and the algorithm contests does not generally have a very strict requirement about time complexity, it can generally be replaced by the ST table, which is not practical in the algorithm competition. Here we offer an improved version that is more practical in the contests.
```

Comment:
> Because the Four Russians algorithm is based on the ST table, and the algorithm contests does not generally have a very strict requirement about time complexity, it can generally be replaced by the ST table, **which is not practical in the algorithm competition**. Here we offer an improved version that is more practical in the contests.
这里表述有点问题，这样说的话 `not practical` 修饰的是 `ST table`